### PR TITLE
Fix false negatives for CSS-In-JS in color-no-invalid-hex #3945

### LINKED ISSUE #3945: Fix false negatives for CSS-In-JS in color-no-invalid-hex
Stylelint does not catch CSS errors in JSX, and according to https://github.com/stylelint/stylelint/issues/3264, out-of-the-box JSX support is expected.

```javascript
// Test.jsx
import React from "react";

export default function Test() {
  return <h1 style={{ color: "#FFFFF" }}>Test</h1>; // Invalid hex code
}
```

```javascript
// .stylelintrc
{
  "extends": ["stylelint-config-recommended"]
}
```

```javascript
// package.json
{
  "scripts": {
    "stylelint": "stylelint"
  },
  "dependencies": {
    "react": "16.8.0"
  },
  "devDependencies": {
    "stylelint": "^9.10.1",
    "stylelint-config-standard": "^18.2.0",
  }
}
```

Then, running `yarn stylelint Test.jsx` has the following output:

```
$ stylelint Test.jsx
✨  Done in 0.84s.
```

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

It is related to JSX syntax.

> What did you expect to happen?

It should lint the style prop.

> What actually happened (e.g. what warnings or errors did you get)?

No linting errors.

### PR DESCRIPTION
**Pull request by fogost**
> Which issue, if any, is this issue related to?

Closes #3945

> Is there anything in the PR that needs further explanation?

This solves the problem, but I'm not sure about tests.

### CODE REVIEW

File: lib/rules/color-no-invalid-hex/__tests__/index.js [hunk 1]
```diff
@@ -28,9 +28,6 @@ testRule(rule, {
     {
       code: "a { padding: 000; }"
     },
-    {
```
Comment (jeddy3) [inline lib/rules/color-no-invalid-hex/__tests__/index.js]: This test needs to stay as we need to prove that this rule doesn't check the contents of strings.

File: lib/rules/color-no-invalid-hex/index.js [hunk 2]
```diff
@@ -23,36 +23,39 @@ const rule = function(actual) {
...
 
-      styleSearch({ source: declString, target: "#" }, match => {
-        // If there's not a colon, comma, or whitespace character before, we'll assume this is
-        // not intended to be a hex color, but is instead something like the
-        // hash in a url() argument
-        if (!/[:,\s]/.test(declString[match.startIndex - 1])) {
-          return;
-        }
+      styleSearch(
+        { source: declString, target: "#", strings: "check" },
```
Comment (jeddy3) [inline lib/rules/color-no-invalid-hex/index.js]: I suspect the way to fix this is to migrate the rule away from `styleSearch` and to `postcss-value-parser`. There wasn't a value parser around when this rule was created, but now we have one and it is proving a more robust way of dealing with the values of declarations.

There an example of the value parser in use in the [color-named rule](https://github.com/stylelint/stylelint/blob/348c139ef1b1c69237a71225d55dfdee73cce39c/lib/rules/color-named/index.js#L73).

e.g.

```js
root.walkDecls(decl => {
  valueParser(decl.value).walk(({value, type, sourceIndex) => {
    if (type !== "word") return;
    const hexMatch = /^#[0-9A-Za-z]+/.exec(value);
    ..
  }
}
```

**Comment by fogost** [comment]
I changed `styleSearch` to `postcss-value-parser` and added css-in-js tests

**Comment by jeddy3** [comment]
> -   Fixed: `color-no-invalid-hex`  false negatives for CSS-in-JS ([#3957](https://github.com/stylelint/stylelint/pull/3957)).

**Pull request merged**